### ISSUE TITLE
nested things get parsed bad

### ISSUE THREAD

**Issue by sils1297** [mkdocs/mkdocs#488]
feel free to edit the title to a better one. I find this hard to describe.

Please compare:
http://coala.readthedocs.org/en/wip-sils-docs/Getting_Involved/Writing_Tests/#actually-writing-a-test
https://github.com/coala-analyzer/coala/blob/master/doc/Getting_Involved/Writing_Tests.md#actually-writing-a-test

In the readthedocs version (you can also clone our repository and run `mkdocs serve` and go to http://127.0.0.1:8000/Getting_Involved/Writing_Tests/#reason-why-this-function-is-untestable ) the `# Reason why this function is untestable.` which is embedded in `>`'s to simulate notes behaves badly.

Note that the github markdown parser seems to handle this well. This seems to apply for other nested constructs too, not only quotes/notes (e.g. code in bullets).

**Comment by d0ugal** [comment]
Hmm, yeah - I see two issues here.

- blockquotes are not that well formatted on the readthedocs theme - we could do with some CSS additons for it probably
- Code isn't being highlighted within a blockquote.

**Comment by facelessuser** [comment]
> blockquotes are not that well formatted on the readthedocs theme - we could do with some CSS additons for it probably

I wouldn't say this is an issue per se as this is readthedocs choosen method of handling blockquotes, but yeah I've thought about styling the blockquotes in a more obvious way.

> Code isn't being highlighted within a blockquote.

This is actually a Python Markdown issue: https://github.com/waylan/Python-Markdown/issues/53.  This is due to how block are handled in Python Markdown, which makes it hard to handle certain kinds of block parsing when nested.  @waylan as mentioned he plans to re-write the block parsers to make it easier to handle these kinds of situations.  But currently, handling these scenarios doesn't work too good.  So what to do until then...

As waylan sounds like he has a clear vision of how he plans on fixing this in the future, I decided to write a Python Markdown extension to workaround the issue until he solves the problem for PythonMarkdown 3. I wrote an extensions which I call [SuperFences](https://github.com/facelessuser/pymdown-extensions/blob/master/pymdownx/superfences.py) that adds support for nesting code blocks (among other things) in other nested blocks (all the blocks including in the Python Markdown default plugin set).  It's from a set of personal extensions I wrote for a command line markdown preview tool, but any of the extensions can be run with mkdocs.

Check out the end result here (nested code blocks are handled in a more reasonable fashion): http://facelessuser.github.io/PyMdown/extensions/superfences/.  

Compared against the source markdown: https://raw.githubusercontent.com/facelessuser/PyMdown/master/docs/extensions/superfences.md.

So if you must have nested blocks, this is a workaround I use until Python Markdown 3 handles these issues better in the future.

You can see I just add my personal extensions to mkdocs in the usual way: https://github.com/facelessuser/PyMdown/blob/master/mkdocs.yml.  Anyways, maybe this can help you in the short term.

**Comment by d0ugal** [comment]
> > blockquotes are not that well formatted on the readthedocs theme - we could do with some CSS additons for it probably

> I wouldn't say this is an issue per se as this is readthedocs choosen method of handling blockquotes, but yeah I've thought about styling the blockquotes in a more obvious way.

Quote often the HTML we generate doesn't work with the ReadTheDocs CSS, so we might need to do some tweaking there. I can't find an example of a block quote and sphinx project to see what it should look like.

I've removed the milestone as it sounds like this isn't going to be something we can fix - I incorrectly assumed this was an issue with out highlighting integration.

**Comment by facelessuser** [comment]
> Quite often the HTML we generate doesn't work with the ReadTheDocs CSS, so we might need to do some tweaking there. I can't find an example of a block quote and sphinx project to see what it should look like.

All they seem to do is indent blocks, I think they may also style `<cite>` tags in a blockquote, but basically, all they do is indent the block...not very clear.  A simple bar on the left would be plenty, or maybe an indent surrounded with quote icons, or anything besides a subtle indent :).  Here's an example of readthedocs and block quotes: http://rest-sphinx-memo.readthedocs.org/en/latest/ReST.html#block-quote.   This is why I will probably augment the readthedocs theme for my personal use.

So basically, the above blockquotes look fine in the opening post (though I think he really should be using admonitions as they are a better fit for what he is doing).  Now their may be more blockquote issues (especially when nesting things like fenced blocks).  But in my experience, all of *my* functional issues were resolved with the workaround posted above, but your milage may vary, and you have to adhere to its specs laid out in the documentation for a reliable experience.

**Comment by d0ugal** [comment]
Okay, that seems reasonable. This needs to be fixed in python-markdown and is out of our scope. Thanks for the detailed input.

**Comment by waylan** [comment]
I haven't looked through this discussion carefully yet, but if I'm understanding things correctly this is about nesting fenced code blocks (no one actually mentions that they are "fenced" code blocks -- which are not standard Markdown). I can confirm that nested fenced code blocks are currently not supported by Python-Markdown. Fenced code blocks can only be at the root of the document. If you want to nest code blocks you **must** use indented code blocks. Of course, that makes it more difficult to get JavaScript libraries to highlight those blocks as there is no way to define the language -- unless you use the [CodeHilite][1] extension.

The CodeHilite extension uses Pygments and highlights the code blocks once at build time, removing the need for a JavaScript Library. Unfortunately, it is currently not easy to disable the JavaScript Library included with mkdocs (see #467). 

However, The CodeHilite Extension also has a `use_pygments` config option, which, when turned off (set to `False`), outputs code blocks formatted the same as fenced code blocks (the language is defined as a class on the `<code`> tag) so that a JavaScript library could highlight the block.

Note that there are at least three requirements for this to work: 

1. You will need to have mkdocs version 0.13 or later, which has not been released yet. The ability to set config options for Python-Markdown Extensions will be introduced in that version when it is released (see #435). If you want it now, you will have to use the development version here on GitHub.
2. You will need to use Python-Markdown version 2.6.x (the latest) as the `use_pygments` config option was introduced in that version (see the [release notes][3]).
3. You will need to use CodeHilite's [method][2] of defining the language for each indented code block.

[1]: https://pythonhosted.org/Markdown/extensions/code_hilite.html
[2]: https://pythonhosted.org/Markdown/extensions/code_hilite.html#syntax
[3]: https://pythonhosted.org/Markdown/release-2.6.html#pygments-can-now-be-disabled

**Comment by facelessuser** [comment]
> I haven't looked through this discussion carefully yet, but if I'm understanding things correctly this is about nesting fenced code blocks (no one actually mentions that they are "fenced" code blocks -- which are not standard Markdown).

True, I guess this is an important point.  And you've listed some good steps to get working nested code blocks within the current default plugin set.  I have found in general using the indented blocks are cumbersome in nested environments, but this is another option.

In general, I think most people are waiting for the day that fenced blocks can be supported in a nested environment despite it not being a "Markdown standard" as they are far less cumbersome to use.  The potential of Python Markdown block processors being reworked (https://github.com/waylan/Python-Markdown/issues/53) in a more sane way will **hopefully** allow this to be much easier to implement as I really had to do some coding gymnastics to get fenced blocks working reliably in nested environments, but I would say now that it is pretty solid (within the plugin subset I use).

If fenced blocks can be processed as actual blocks instead of as a preprocesser, this will allow detecting the nesting even better.  The current block implementation breaks down due to the way lists are processed and fed to the other block processes.
